### PR TITLE
[SofaCore & runSofa] Handle optional data field with default value.

### PR DESCRIPTION
**Pull request by damienmarchal**
This is the follow up of discussion: https://github.com/sofa-framework/sofa/issues/1675

The UI part is not very important but the change introduced in BaseData.h and BaseObject deserve 
a bit of attention. I will pass it to ready when I will need for a review but If you have some time @hugtalbot I would like to have 
your opinion on the change in BaseData.

**EDIT**:   

This PR allows to define optional datafields. Here's how it looks. 
If a datafield is set to have no default value (using the new `setWithDefaultValue()`), the gui will display `<no default value>`. 

Additionally, if a datafield is required and has no default value, an error will show up on the terminal and the gui will display `<missing value>`.  

Screenshot from 2021-08-03 11-31-22

______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

### CODE REVIEW

**Comment by damienmarchal** [comment]
@EulalieCoevoet Can you have a look at that ? It is WIP.

File: SofaKernel/modules/SofaCore/src/sofa/core/objectmodel/BaseObject.cpp [hunk 1]
```diff
@@ -254,9 +254,19 @@ void BaseObject::init()
...
+            {
+                if(data->hasDefaultValue())
+                {
+                    msg_info() << "Required data \"" << data->getName() << "\" has not been set but has a default value. (Current value is " << data->getValueString() << ")" ;
+                }
+                else
+                {
+                    msg_error() << "Required data \"" << data->getName() << "\" has not been set and has not default value. ";
+                }
+            }
```
Comment (EulalieCoevoet) [inline SofaKernel/modules/SofaCore/src/sofa/core/objectmodel/BaseObject.cpp]: If isRequired and hasDefaultValue are set to true by default, this is a bit extreme in terms of verbosity... 
I agree with this approach, but now I think isRequired() should be false by default. There are much more optional data than required ones.

Comment (EulalieCoevoet) [reply SofaKernel/modules/SofaCore/src/sofa/core/objectmodel/BaseObject.cpp]: @damienmarchal what do you think?

Comment (damienmarchal) [reply SofaKernel/modules/SofaCore/src/sofa/core/objectmodel/BaseObject.cpp]: Ouch we are already March 1.

**Comment by guparan** [comment]
@damienmarchal This PR has been inactive and WIP for 6 months. 
Do you plan working on it before September? If no, let's close+postpone it.

**Comment by EulalieCoevoet** [comment]
Arf we were so close! I think it just needs a bit of cleaning now, or at least that's what I remember from our last coding session. 
@damienmarchal am I wrong? 
If so I can do it.

**Comment by EulalieCoevoet** [comment]
I'd like this PR to be reviewed as I think the new features are really nice, especially for the SoftRobots plugin. 
@damienmarchal would you agree to remove the WIP label?

**Pull request closed without merging**